File: cpp/mainwindow.cpp
#include "mainwindow.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    QMenuBar *menubar = menuBar();

    QMenu *menu = menubar->addMenu(tr("File"));
    auto action = new QAction(tr("Quit"), this);
    action->setIcon(QIcon::fromTheme("application-exit"));
    connect(action, &QAction::triggered, qApp, &QApplication::quit);

    menu->addAction(action);
    ui->mainToolBar->addAction(action);
}


MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::loadLocation()
{
    QUrl url = ui->lineEdit->text();
    if (url.scheme().isEmpty())
        url.setScheme("https");
    ui->webView->load(url);
}

void MainWindow::urlChanged(const QUrl &url)
{
    ui->lineEdit->setText(url.toString());
}

File: python/thebrowser/mainwindow.py
import sys
import os.path
from PyQt5 import uic
from PyQt5.QtCore import QUrl
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QAction, qApp

uifile = os.path.dirname(__file__) + "/../../cpp/mainwindow.ui"
MainWindowUI, MainWindowBase = uic.loadUiType(uifile)

class MainWindow(MainWindowBase, MainWindowUI):
    def __init__(self, parent=None):
        super().__init__()
        self.setupUi(self)

        menubar = self.menuBar
        menu = menubar.addMenu(self.tr("File"))
        action = QAction(self.tr("Quit"), self)
        action.setIcon(QIcon.fromTheme("application-exit"))
        action.triggered.connect(qApp.quit)

        menu.addAction(action)
        self.mainToolBar.addAction(action)

    def loadLocation(self):
        url = QUrl(self.lineEdit.text())
        if not url.scheme():
            url.setScheme("https")
        self.webView.load(url)

    def urlChanged(self, url):
        self.lineEdit.setText(url.toString())


